pgmpy/inference/ExactInference.py

```diff
@@ -171,24 +172,25 @@ def _variable_elimination(
             # Removing all the factors containing the variables which are
             # eliminated (as all the factors should be considered only once)
             factors = [
-                factor
+                factor[0]
                 for factor in working_factors[var]
-                if not set(factor.variables).intersection(eliminated_variables)
+                if not set(factor[0].variables).intersection(eliminated_variables)
             ]
             phi = factor_product(*factors)
             phi = getattr(phi, operation)([var], inplace=False)
             del working_factors[var]
             for variable in phi.variables:
-                working_factors[variable].add(phi)
+                working_factors[variable].add((phi, var))
             eliminated_variables.add(var)
 
         # Step 4: Prepare variables to be returned.
         final_distribution = set()
         for node in working_factors:
-            factors = working_factors[node]
-            for factor in factors:
-                if not set(factor.variables).intersection(eliminated_variables):
+            for factor in working_factors[node]:
```

Comment:
@rballester I would rather do something like `for factor, _ in working_factors[node]`. Would not need the next line then.

Reply:
In this case both elements of the tuple, not just the first one, are needed for line 192 (final_distribution.add(factor)).

We could do this if you prefer it:
```suggestion
            for factor, origin in working_factors[node]:
                if not set(factor.variables).intersection(eliminated_variables):
                    final_distribution.add((factor, origin))
```

Reply:
Yes, this should work.